citre-lang-c.el

```diff
@@ -0,0 +1,381 @@
+;;; citre-lang-c.el --- Language support for C -*- lexical-binding: t -*-
+
+;; Copyright (C) 2021 Hao WANG
+
+;; Author: Hao WANG <amaikinono@gmail.com>
+;; Maintainer: Hao WANG <amaikinono@gmail.com>
+;; Created: 28 Jan 2021
+;; Keywords: convenience, tools
+;; Homepage: https://github.com/AmaiKinono/citre
+;; Version: 0
+
+;; This file is NOT part of GNU Emacs.
+
+;; This program is free software; you can redistribute it and/or
+;; modify it under the terms of the GNU General Public License
+;; as published by the Free Software Foundation; either version 3
+;; of the License, or (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Language support for C.
+;;
+;; The main feature is file reachability analysis:
+;;
+;; 1. It's only done when the symbol is not in top-scope, since they may be
+;; declarations of extern symbols.  An extern "C" block is not count as a
+;; scope.
+;;
+;; 2. We find all header files included in current file.  This is done
+;; recursively, so headers included by headers are also recorded.  When we find
+;; a macro in include directive, we give up the analysis.
+;;
+;; 3. When the analysis is successfully done, we only keep tags that's
+;; reachable, which means it's from current file, or a file relevant to one of
+;; the headers.
+;;
+;; For example, if a header "module.h" is included.  By "relevant to it", we
+;; mean the file path contains "module" as a directory or file name.  The
+;; reason for doing it this way, rather than requiring the file name to be
+;; "module.h" or "module.c", is to support multi-language project.  C is a
+;; widely used language for interop, so we need tags from other languages too.
+;; It's pretty safe to assume that a module written in another language would
+;; have "module" in its path, if it's used by including "module.h".  Also, this
+;; actually works very well for pure C projects.
+;;
+;; Other noticeable things are:
+;;
+;; - When the symbol is after "->" or ".", tags of member kind are sorted above
+;;   others.
+;; - When the symbol is before "(", tags of function/macro kinds are sorted
+;;   above others
```

Comment:
Really nice. If the all characters in the name in a token at the left side of '(' is upper cased, macro kind are sorted above function kind.

Reply:
After some thinking I think we shouldn't do this. If the project has the naming convention of using uppercase for macros, then macros and functions won't appear at the same time. The fact that the symbol is upper cased can make sure there are only macros in the result. If the project doesn't have such convention, then we are doing unnecessary things.